File: midonet-api/src/main/java/org/midonet/api/neutron/loadbalancer/VipResource.java
/*
 * Copyright 2014 Midokura SARL
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.midonet.api.neutron.loadbalancer;

import java.util.List;
import java.util.UUID;

import javax.annotation.security.RolesAllowed;
import javax.ws.rs.Consumes;
import javax.ws.rs.DELETE;
import javax.ws.rs.GET;
import javax.ws.rs.POST;
import javax.ws.rs.PUT;
import javax.ws.rs.Path;
import javax.ws.rs.PathParam;
import javax.ws.rs.Produces;
import javax.ws.rs.core.Response;
import javax.ws.rs.core.SecurityContext;
import javax.ws.rs.core.UriInfo;

import com.google.inject.Inject;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import org.midonet.cluster.auth.AuthRole;
import org.midonet.api.rest_api.AbstractResource;
import org.midonet.api.rest_api.RestApiConfig;
import org.midonet.cluster.rest_api.NotFoundHttpException;
import org.midonet.cluster.rest_api.neutron.NeutronMediaType;
import org.midonet.cluster.rest_api.neutron.models.VIP;
import org.midonet.cluster.services.rest_api.neutron.plugin.LoadBalancerApi;
import org.midonet.event.neutron.VipEvent;

import static org.midonet.cluster.rest_api.validation.MessageProperty.RESOURCE_NOT_FOUND;
import static org.midonet.cluster.rest_api.validation.MessageProperty.getMessage;

public class VipResource extends AbstractResource {

    private final static Logger log = LoggerFactory.getLogger(
        VipResource.class);
    private final static VipEvent VIP_EVENT = new VipEvent();

    private final LoadBalancerApi api;

    @Inject
    public VipResource(RestApiConfig config, UriInfo uriInfo,
                       SecurityContext context, LoadBalancerApi api) {
        super(config, uriInfo, context, null, null);
        this.api = api;
    }

    @GET
    @Path("{id}")
    @Produces(NeutronMediaType.VIP_JSON_V1)
    @RolesAllowed(AuthRole.ADMIN)
    public VIP get(@PathParam("id") UUID id) {
        log.info("VipResource.get entered {}", id);

        VIP vip = api.getVip(id);
        if (vip == null) {
            throw new NotFoundHttpException(getMessage(RESOURCE_NOT_FOUND));
        }

        log.info("VipResource.get exiting {}", vip);
        return vip;
    }

    @GET
    @Produces(NeutronMediaType.VIPS_JSON_V1)
    @RolesAllowed(AuthRole.ADMIN)
    public List<VIP> list() {
        log.info("VipResource.list entered");
        return api.getVips();
    }

    @POST
    @Consumes(NeutronMediaType.VIP_JSON_V1)
    @Produces(NeutronMediaType.VIP_JSON_V1)
    @RolesAllowed(AuthRole.ADMIN)
    public Response create(VIP vip) {
        log.info("VipResource.create entered {}", vip);

        api.createVip(vip);
        VIP_EVENT.create(vip.id, vip);
        log.info("VipResource.create exiting {}", vip);
        return Response.created(
            LBUriBuilder.getVip(getBaseUri(), vip.id))
            .entity(vip).build();
    }

    @DELETE
    @Path("{id}")
    @RolesAllowed(AuthRole.ADMIN)
    public void delete(@PathParam("id") UUID id) {
        log.info("VipResource.delete entered {}", id);
        api.deleteVip(id);
        VIP_EVENT.delete(id);
    }

    @PUT
    @Path("{id}")
    @Consumes(NeutronMediaType.VIP_JSON_V1)
    @Produces(NeutronMediaType.VIP_JSON_V1)
    @RolesAllowed(AuthRole.ADMIN)
    public Response update(@PathParam("id") UUID id, VIP vip) {
        log.info("VipResource.update entered {}", vip);

        api.updateVip(id, vip);
        VIP_EVENT.update(id, vip);
        log.info("VipResource.update exiting {}", vip);
        return Response.ok(
            LBUriBuilder.getVip(getBaseUri(), vip.id))
            .entity(vip).build();
    }
}

File: midonet-cluster/src/main/scala/org/midonet/cluster/services/rest_api/resources/VipResource.scala
/*
 * Copyright 2015 Midokura SARL
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.midonet.cluster.services.rest_api.resources

import java.util.{List => JList, UUID}

import javax.ws.rs.core.MediaType.APPLICATION_JSON
import javax.ws.rs.core.Response
import javax.ws.rs.core.Response.Status
import javax.ws.rs.{HeaderParam, POST, Path, WebApplicationException, _}

import scala.collection.JavaConverters._

import com.google.inject.Inject
import com.google.inject.servlet.RequestScoped

import org.midonet.cluster.rest_api.Status.METHOD_NOT_ALLOWED
import org.midonet.cluster.rest_api.annotation._
import org.midonet.cluster.rest_api.models.{LoadBalancer, Pool, Vip}
import org.midonet.cluster.rest_api.{BadRequestHttpException, NotFoundHttpException}
import org.midonet.cluster.services.rest_api.MidonetMediaTypes._
import org.midonet.cluster.services.rest_api.resources.MidonetResource.{NoOps, Ops, ResourceContext}

@ApiResource(version = 1, name = "vips", template = "vipTemplate")
@Path("vips")
@RequestScoped
@AllowCreate(Array(APPLICATION_VIP_JSON,
                   APPLICATION_JSON))
@AllowUpdate(Array(APPLICATION_VIP_JSON,
                   APPLICATION_JSON))
class VipResource @Inject()(resContext: ResourceContext)
    extends MidonetResource[Vip](resContext) {

    @GET
    @Path("{id}")
    @Produces(Array(APPLICATION_VIP_JSON,
                    APPLICATION_JSON))
    override def get(@PathParam("id") id: String,
                     @HeaderParam("Accept") accept: String):Vip = {
        val vip = getResource(classOf[Vip], id).getOrThrow
        val pool = getResource(classOf[Pool], vip.poolId).getOrThrow
        vip.loadBalancerId = pool.loadBalancerId
        vip
    }

    @GET
    @Produces(Array(APPLICATION_VIP_COLLECTION_JSON,
                    APPLICATION_JSON))
    override def list(@HeaderParam("Accept") accept: String): JList[Vip] = {
        val vips = listResources(classOf[Vip]).getOrThrow
        for (vip <- vips) {
            vip.loadBalancerId =
                getResource(classOf[Pool], vip.poolId).getOrThrow.loadBalancerId
        }
        vips.asJava
    }

    @DELETE
    @Path("{id}")
    override def delete(@PathParam("id") id: String): Response = {
        try {
            val response = super.delete(id)
            if (response.getStatus == Status.NOT_FOUND.getStatusCode)
                MidonetResource.OkNoContentResponse
            else
                response
        } catch {
            case e: WebApplicationException
                if e.getResponse.getStatus == Status.NOT_FOUND.getStatusCode =>
                MidonetResource.OkNoContentResponse
        }
    }

    protected override def createFilter(vip: Vip): Ops = {
        vip.create()
        throwIfViolationsOn(vip)
        try {
            getResource(classOf[Pool], vip.poolId).getOrThrow
        } catch {
            case t: NotFoundHttpException =>
                throw new BadRequestHttpException(t.getMessage)
        }
        NoOps
    }

    protected override def updateFilter(to: Vip, from: Vip): Ops = {
        to.update(from)
        try {
            getResource(classOf[Pool], to.poolId).getOrThrow
        } catch {
            case t: NotFoundHttpException =>
                throw new BadRequestHttpException(t.getMessage)
        }
        NoOps
    }

}

@RequestScoped
@AllowCreate(Array(APPLICATION_VIP_JSON,
                   APPLICATION_JSON))
class PoolVipResource @Inject()(poolId: UUID, resContext: ResourceContext)
    extends MidonetResource[Vip](resContext) {

    @GET
    @Produces(Array(APPLICATION_VIP_COLLECTION_JSON,
                    APPLICATION_JSON))
    override def list(@HeaderParam("Accept") accept: String): JList[Vip] = {
        getResource(classOf[Pool], poolId) flatMap { pool =>
            listResources(classOf[Vip], pool.vipIds.asScala) map { vips =>
                vips.foreach(_.loadBalancerId = pool.loadBalancerId)
                vips.asJava
            }
        } getOrThrow
    }

    protected override def createFilter(vip: Vip): Ops = {
        vip.create()
        throwIfViolationsOn(vip)
        val pool = try {
            getResource(classOf[Pool], vip.poolId).getOrThrow
        } catch {
            case t: NotFoundHttpException =>
                throw new BadRequestHttpException(t.getMessage)
        }
        vip.loadBalancerId = pool.loadBalancerId
        NoOps
    }
}

@RequestScoped
class LoadBalancerVipResource @Inject()(loadBalancerId: UUID,
                                        resContext: ResourceContext)
    extends MidonetResource[Vip](resContext) {

    @GET
    @Produces(Array(APPLICATION_VIP_COLLECTION_JSON,
                    APPLICATION_JSON))
    override def list(@HeaderParam("Accept") accept: String): JList[Vip] = {
        getResource(classOf[LoadBalancer], loadBalancerId) flatMap { lb =>
            listResources(classOf[Pool], lb.poolIds.asScala)
        } flatMap { pools =>
            val vipIds = pools.flatMap(_.vipIds.asScala)
            listResources(classOf[Vip], vipIds) map { vips =>
                vips.foreach(_.loadBalancerId = loadBalancerId)
                vips.asJava
            }
        } getOrThrow
    }

    @POST
    override def create(v: Vip, @HeaderParam("Content-Type") cType: String)
    : Response = {
        Response.status(METHOD_NOT_ALLOWED.getStatusCode).build()
    }

}